prov/util/src/util_mr_cache.c

```diff
@@ -310,6 +311,17 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *att
 
 		cache->search_cnt++;
 		*entry = cache->storage.find(&cache->storage, &info);
+
+		if (*entry &&
+		    !ofi_monitor_valid(cache->monitor,
+				       (const void *)(*entry)->info.iov.iov_base,
+				       (*entry)->info.iov.iov_len)) {
+			ofi_monitor_unsubscribe(cache->monitor,
```

Comment:
Do you need this unsubscribe? The entry is going to get purged if it is not usable anyway.

Reply:
The unsubscribe will purge the entry from the CUDA memory monitor. Since the CUDA memory monitor does not actively check regions, it needs the MR cache to purge entries. Or, entries will be purged if a threshold is exceeded.

Reply:
There may be multiple entries in the cache that are covered by the new region, all of which need to be removed in order that future searches will work.  Try to match the behavior of the original code -- check first for a usable entry and goto hit in that case.  If the entry is not usable, we need the while loop below to purge all unusable entries.

Reply:
I'll work on cleaning this up. One thing that may not be clear at first glance is that the CUDA memory monitor monitors an entire user allocated region. When ofi_monitor_unsubscribe() is called, the CUDA memory monitor will free the corresponding monitored region and call ofi_monitor_notify() for the entire region at the end of the free (ofi_cuda_mm_entry_free()). I believe this should cleanup all stale entries in all subscribed MR caches. But, this back and forth between the MR cache and CUDA memory monitor may not be ideal.

Reply:
The entries that a new region may match may be a mix of device and host buffers.

Reply:
Will fix.